codeU2.py

```diff
@@ -0,0 +1,181 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jan 13 10:22:28 2017
+Exercise set 2, Code U
+@author: Luiza
+"""
+import logging
+import unittest
+
+class Tree():
+    
+    """Defines the class for the tree:
+    self.children contains a dict of pairs (node:(left_child,right_child))
+    self.head contains the value of the tree head
+    None denotes the absence if the node
+    """
+    
+    def __init__(self,children,head):
+        self.children = children;
+        self.head = head;
+
+def dfs_path(tree_, start, end, path=None):
+
+    """Yields all full paths from start node
+    to end node, excluding the end node
+    """
+    
+    tree = tree_.children
+    stack = [(start, [start])]
```

Comment:
I'd like to see a little comment explaining the algorithm implemented here. It looks like descending depth first into the tree breaking the descent (and trying other branches) when the `end` node is found?

The algorithm explicitly allows for `end` values in the tree (by not breaking on the first path found) but then the set arithmetic `set(tree[vertex]) - set(path)` looks inconsistent in not allowing multiples of any intermediate nodes.

Is that to not get into loops? One would imagine that loops should not be allowed (it wouldn't be a tree if they were). Maybe it would be better to check for loops in the constructor. For the purposes of this assignment, it would probably be enough to just document that "trees" with loops are forbidden and leave the actual check out.

In any case, it took me a bit of time to trace what this function is doing and I suspect it took you a bit of careful thinking to write it as well. Having to carefully think about your code is a great signal for you to also write some comments to the code so the future readers can skip the "figuring out" part and get the conclusions from the helpful comments.